File: scala/src/test/scala/com/shekhargulati/ninetynine_problems/_01_lists/P12Test.scala
package com.shekhargulati.ninetynine_problems._01_lists

import org.scalatest.{FunSpec, Matchers}

class P12Test extends FunSpec with Matchers {

  describe("decode spec") {
    it("should decode an run-length encoded list") {
      val decoded: List[String] = P12.decode(List((4, "a"), "b", (2, "c"), (2, "a"), "d", (4, "e")))
      decoded should have length 14
      decoded should equal(List("a", "a", "a", "a", "b", "c", "c", "a", "a", "d", "e", "e", "e", "e"))
    }
  }

}

File: java8/src/test/java/com/shekhargulati/ninetynine_problems/_01_lists/P12Test.java
package com.shekhargulati.ninetynine_problems._01_lists;

import org.junit.Test;

import java.util.AbstractMap.SimpleEntry;
import java.util.Arrays;
import java.util.List;

import static org.hamcrest.collection.IsCollectionWithSize.hasSize;
import static org.junit.Assert.assertThat;

public class P12Test {

    @Test
    public void shouldDecodeEncodedList() throws Exception {
        List<String> encoded = P12.decode(
                Arrays.asList(
                        new SimpleEntry<>(4, "a"),
                        "b",
                        new SimpleEntry<>(2, "c"),
                        new SimpleEntry<>(2, "a"),
                        "d",
                        new SimpleEntry<>(4, "e")));

        assertThat(encoded, hasSize(14));
    }
}